### ISSUE TITLE
Zettlr does nothing when file is modified by another app

### ISSUE THREAD

**Issue by Saya47** [Zettlr/Zettlr#4626]
### Description

Hello when I open a markdown file in Zettlr, and say while it's open on Zettlr, it gets modified by another app such as Sublime, the changes don't show up on Zettrl. Even when I close the file from the editor and reopen it, Zettlr shows me the old version of the file. I literally have to close the file, close Zettlr and reopen it and reopen the file to see the updates.

### Proposed Changes

In text editors such as Notepad++ when a file gets changed by a third party, a prompt is displayed to the user to keep the current version of file or show the changes.
In other editors such as VS Code, when a file gets changes, the changes are immediately fetched from file and shown to the user.

### Caveats

None.Let's say I have a lot of markdown notes, and I edit them because they are documentations to software etc. And I need different editors to manage and spread my work. And if I accidentally changed a markdown file with a third party then edited the same file that happened to be previously opened by Zettlr, and do modifications on Zettlr, I will lose the modifications that I have done with the other app. So this is very important to me.

### Do you Wish to Attempt Implementing this Yourself?

No

### Zettlr Version

Stable (most recent version)

### Your Platform

- [X] Windows
- [ ] macOS
- [ ] Linux

### Operating System Version

Windows 10 1803

**Comment by kyaso** [comment]
There is an option to load remote changes into a file:
![image](https://github.com/Zettlr/Zettlr/assets/18646029/6af7b16a-4e74-4060-ac03-9abbffc3a51d)

Could you try with that option enabled and see if it works?

**Comment by Saya47** [comment]
Oh I didn't know that @kyaso but it seems this option is enabled by default. So maybe I'm experiencing a bug here. I will do further examinations and report back. Thank you.
![image](https://github.com/Zettlr/Zettlr/assets/35542834/603602e4-63cd-405e-b901-d64402958e5f)

**Comment by Saya47** [comment]
Issue still persists

**Comment by BabakFarshchian** [comment]
I noticed the same problem. Change to a file in another application, Zettlr does not notice. I close the file, even restart Zettlr without effect. Only when I clear FSAL cache (see the help menu) do changes show up in Zettlr. So I guess (at least from my end) this seems to be one of the issues related to FSAL cache.

**Comment by Gerigory** [comment]
> I noticed the same problem. Change to a file in another application, Zettlr does not notice. I close the file, even restart Zettlr without effect. Only when I clear FSAL cache (see the help menu) do changes show up in Zettlr. So I guess (at least from my end) this seems to be one of the issues related to FSAL cache.

Clear FSAL cache won't help either

**Comment by nathanlesage** [comment]
I recently found out that, if you open a file and then close it, Zettlr won't react to any changes. BUT if you keep it open as the changes happen, it will suddenly do what it's supposed to. There's a related issue that gave me this idea.

**Comment by robert-smith-maersk** [comment]
I had 'load remote changes' enabled. For me (Zettlr 4.1.1 installed using Homebrew on Mac OS) that didn't work. Enabling 'Watchdog Polling' in advanced settings produces the expected behaviour.

**Comment by nathanlesage** [comment]
> I had 'load remote changes' enabled. For me (Zettlr 4.1.1 installed using Homebrew on Mac OS) that didn't work. Enabling 'Watchdog Polling' in advanced settings produces the expected behaviour.

This really does indicate an issue with chokidar. This one has been updated to a major release before our 4.1.1 release, so that seems to be the culprit.

I'll check that.

**Comment by nathanlesage** [comment]
So, on macOS (Silicon), using the regular 4.1.1 release, changes to my files are indeed being detected. Odd.

**Comment by dxpalma** [comment]
I confirm the issue on version 4.1.1 / Debian 13, and that enabling the watchdog worksaround it.

Edit : However, enabling the watchdog does not solves the issue of internal links update after a file rename (https://github.com/Zettlr/Zettlr/issues/4989). The file must be closed and reopened for the link to be updated.

**Comment by nathanlesage** [comment]
To your edit-comment: that is going to be very hard. When you rename a file within Zettlr, we can see that a file is being renamed. However, when the renaming occurs externally, Zettlr sees two independent events -- one file is deleted and another is created. That means it will be very hard to relate a deletion to a creation in such a way as to figure out if the newly appearing file is one that just disappeared.

**Comment by dxpalma** [comment]
My apalogies, I wasn't precise enough :
I was talking about file renaming within Zettlr. I didn't even considered external renaming, well, because as you said, this would be very difficult to detect.
This is likely to be a cache or buffer update issue :
* Works : Close all tabs and the app. Open the app and open/rename (Ctrl+r) a page to which a link in another page directs. Validate the internal link update question. Open the other page containing the link : it is correctly updated.
* Doesn't work : Do the same, but prior to renaming the first page, open the other page containing the link. After the renaming, the link isn't updated.

**Comment by dxpalma** [comment]
I just found the solution for my case (which doesn't mean it's the same as @Saya47).
I got a hint when I realized I had the same problem in my file manager than in Zettlr.

It was a problem of fs.inotify.max_user_watches reached.
The solution is to increase its value.

There might be an exception at some point when Zettlr tries to set the watch. It would be nice to have a message when this happens.

On Debian, as root :

Get the current value :
~~~
sysctl fs.inotify.max_user_watches
~~~

Test another value (not reboot proof) :
~~~
sysctl -w fs.inotify.max_user_watches=NEW_VALUE
~~~

Make it permanent for reboot :
~~~
echo "fs.inotify.max_user_watches=NEW_VALUE" >> /etc/sysctl.d/99-custom-max-user-watches.conf
~~~

**Comment by benniekiss** [comment]
> It was a problem of fs.inotify.max_user_watches reached.
The solution is to increase its value.

This is interesting to me, because I've noticed that in large workspaces, zettlr also reaches file watcher limits through chokidar(i think ??), but from what I recall in the logs, its multiple watchers on the same file. I'll see if I can reproduce the errors I was seeing some time today or tomorrow. This might point to issues users have been having with watchdog events, etc

**Comment by farvardin** [comment]
I've experienced the same. What is strange is the first time if I open the file from an other editor and change a letter from there, it will reload correctly in zettlr, but if I do it a second time, it won't. I've tried to increase fs.inotify.max_user_watches value but it changed nothing. Also the watchdog polling change didn't work. 

But if I close ALL other files/tabs, then it works again.

**Comment by fiveangle** [comment]
I just experienced the *worse* sitch, where I made changes to the md file in Zettlre, my agent reverted to a previous branch, and Zettlr just reverted the last hour of changes I made to the file and I'm now struggling to figure out if there is any way to get my changes back.

So far, not blowing my skirt up, if you know what i mean 🥵

Zettlr Version: 4.3.0
Build number: fef1064 (Saturday, March 21, 2026)
UUID: 3632ae28-ef05-4177-ac17-32ea2ff4f804
System: darwin 26.4.0 (architecture: arm64)


EDIT: Yeah, my 2 pages of handwritten text are gone with the wind. F' this… back to Sublime Text…

**Comment by nathanlesage** [comment]
> I just experienced the worse sitch, where I made changes to the md file in Zettlre, my agent reverted to a previous branch, and Zettlr just reverted the last hour of changes I made to the file and I'm now struggling to figure out if there is any way to get my changes back.

Just so I understand what has happened here:

0. You are working in a git repository.
1. You wrote some text into a Markdown file located in that repository using Zettlr.
2. Then you switched to another app (a GPT based one, based on that you mention "agent"?)
3. That app decided to checkout to a different branch and clean up the working state
4. Zettlr recognized that and reloaded the now again empty file

Is this an accurate description of what happened?

**Comment by fiveangle** [comment]
I applaud you for your inquiry. That says something (honestly). And apologize for my frustrated post above, but I am just a human.

Chain of events:

1) I opened an existing README.md file in a checked out git repo and added a comprehensive section to it over 2hrs, then stepped away.
2) Another app (in this case, Opencode running with Big Pickle) reverted the git repo to a different branch as part of its workflow to resolve an issue it was attempting to address.
3) Upon returning from stepping away from my editing session of README.md in Zettlr, I found the contents of the editor window reverted to the same as that when I started the editing session (which also happened to be the then-current on-disk contents of the file in the switched branch of the git repo).

During the Zettlr install wizard, I chose the option to save immediately and view as WYSIWYG. Otherwise, defaults options should have been present.

I do not know why the editing session was reverted to the contents of the README.md as it was prior to the start of my editing session. I assumed it was due to the git checkout by Opencode working in the same dir changing the on-disk contents, but that is just speculation. I checked for any stashes with my changes (since I had enabled the immediate save option in the Zettlr setup wizard) but found none.

For most realtime-cognizant editors (Sublime Text, for example) if the on-disk file is changed during a session, a dialog is presented to the user alerting them of the file being changed on-disk, with option to reload.  If the on-disk contents of the open file changing is what prompted Zettlr to revert to that version, I received no such question to approve that action.

**Comment by nathanlesage** [comment]
Thanks for confirming. First, I am honestly sorry that this has happened to you, but every component was working as expected. When a file is modified remotely, Zettlr correctly in your case picked this up. The mistake was to change branches without committing the changes to your README file first. `git` has correctly overwritten the new and saved, but uncommitted changes with the other state, which contained the README before your changes. Zettlr then correctly identified that the user (that's you) has changed the file contents remotely, and reloaded the now new "correct" state. It cannot know that you did not intend to throw away the changes.

There are three additional notes I have here:

1. When working with git repositories, *always commit your changes*. It's very easy to get any work overridden accidentally in this context.
2. That being said, I am wondering why git did not emit any warning. Can it be that your other app supplied the `--force`-option, which suppresses any warnings that you are about to override changes?
3. Zettlr indeed features an option that prevents it from automatically reading in remote changes, which you can use: [Always Reload Changes](https://docs.zettlr.com/en/reference/settings/#general)

**Comment by fiveangle** [comment]
Thank you for confirming my experience is expected based on the defaults being B.A.D. (broken, as designed, as we call it in my corner of the industry) and not some much scarier ghost in the machine.

I highly urge the team to reconsider the "yolo" out-of-the-box defaults. The "you're holding it wrong" ideology here seems quite misplaced, as Zettlr is an editor-first and I can't imagine any universe where having the contents of an editor be jettisoned without warning as the default, being a reasonable and expected thing.

But I do appreciate you reviewing my situation so quickly and thoroughly.

**Comment by marc-medley** [comment]
`load remote changes` was not operational. 

then, as I was writing up the details I found that _toggling the setting_ `Always load remote changes to the current file` _produced the expected behavior_ for both setting values.

that said, i did not isolate to the mechanism of why, or whether the issue can/will occur again.

tl;dr;

## Initial State: 

**Issue:**

Zettlr did not update (or prompt to update) a currently open markdown file when that file had been externally modified.

**Configuration:**

- Platform: macOS 26.5.2 Tahoe
- Applications: BBEdit version 16.0.2 (16A78, Apple Silicon), MarKEdit Version 1.33.1 (93), Zettlr version 4.6.0 (4.6.0)
- Given file: `example.md`
- `Zettlr` > `Preferences…`
    - `General` > `Behavior` > `Always load remote changes to the current file` : checked [yes]

```md
## Example Header

some other text.
```

**Steps To Reproduce:**

1. Open `example.md` in BBEdit, MarkEdit and Zettlr
2. In BBEdit, modify `## Example Header` to be `## Example Header (modified in BBEdit)` and save the file changes.
3. Observations: 
    - MarkEdit and MarkEdit-preview both update. (Note: verifies file modified event exists)
    - Zettlr files to update

## Continued Checks: 

**Steps Which Resolved Issue … for now:**

1. Set `Always load remote changes to the current file` : unchecked [OFF]
2. Quit and relaunch Zettlr.
3. Repeat **Steps To Reproduce**
4. Observation: Zettlr prompts with "example.md changed on disk …" 
5. Set `Always load remote changes to the current file` : checked [ON]
6. Quit and relaunch Zettlr.
7. Repeat **Steps To Reproduce**
8. Observation: Zettlr automatically updated the changes to "example.md"

**Comment by marc-medley** [comment]
## Continued Use Update:

> The "Zettlr did not update (or prompt to update) a currently open markdown file when that file had been externally modified."

… _issue returned with continued use in a mutli-editor-app context_.

It's as if the setting `Always load remote changes to the current file` somehow ends up forgotten|unset or becomes ignored.

It's not clear how to (easily) get more visibility, or fault isolate, this phenomenon.

**Comment by nathanlesage** [comment]
> It's not clear how to (easily) get more visibility, or fault isolate, this phenomenon.

The setting is not being forgotten. As I have written further up: the culprit is likely chokidar not properly propagating change events to the app, so the app does not get notice of any external changes once the bug in chokidar triggers.

Zettlr relies on chokidar to get notice of any external changes. If that library does not emit the necessary events, there's nothing Zettlr could do.
